src/txrequest.cpp

```diff
@@ -0,0 +1,722 @@
+            // announcement in the ByTxHash index.
+            if (it_prev->m_txhash == it->m_txhash && it_prev->m_state == State::CANDIDATE_READY) {
+                // If one such CANDIDATE_READY exists (for this txhash), convert it to CANDIDATE_BEST.
+                Modify<ByTxHash>(it_prev, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+            }
+        }
+        Modify<ByTxHash>(it, [new_state](Announcement& ann){ ann.m_state = new_state; });
+    }
+
+    //! Check if 'it' is the only announcement for a given txhash that isn't COMPLETED.
+    bool IsOnlyNonCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+        assert(it->m_state != State::COMPLETED); // Not allowed to call this on COMPLETED announcements.
+
+        // This announcement has a predecessor that belongs to the same txhash. Due to ordering, and the
+        // fact that 'it' is not COMPLETED, its predecessor cannot be COMPLETED here.
+        if (it != m_index.get<ByTxHash>().begin() && std::prev(it)->m_txhash == it->m_txhash) return false;
+
+        // This announcement has a successor that belongs to the same txhash, and is not COMPLETED.
+        if (std::next(it) != m_index.get<ByTxHash>().end() && std::next(it)->m_txhash == it->m_txhash &&
+            std::next(it)->m_state != State::COMPLETED) return false;
+
+        return true;
+    }
+
+    /** Convert any announcement to a COMPLETED one. If there are no non-COMPLETED announcements left for this
+     *  txhash, they are deleted. If this was a REQUESTED announcement, and there are other CANDIDATEs left, the
+     *  best one is made CANDIDATE_BEST. Returns whether the announcement still exists. */
+    bool MakeCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+
+        // Nothing to be done if it's already COMPLETED.
+        if (it->m_state == State::COMPLETED) return true;
+
+        if (IsOnlyNonCompleted(it)) {
+            // This is the last non-COMPLETED announcement for this txhash. Delete all.
+            uint256 txhash = it->m_txhash;
+            do {
+                it = Erase<ByTxHash>(it);
+            } while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash);
+            return false;
+        }
+
+        // Mark the announcement COMPLETED, and select the next best announcement (the first CANDIDATE_READY) if
+        // needed.
+        ChangeAndReselect(it, State::COMPLETED);
+
+        return true;
+    }
+
+    //! Move our clock forward and make the data structure consistent with the new time.
+    //! - REQUESTED annoucements with expiry <= now are turned into COMPLETED.
+    //! - CANDIDATE_DELAYED announcements with reqtime <= now are turned into CANDIDATE_{READY,BEST}.
+    void SetTimePoint(std::chrono::microseconds now, std::vector<std::pair<NodeId, GenTxid>>* expired)
```

Comment:
I agree, and considered that, but it's a pain to make that have "any order of calls has well-specified behavior" that introduced a while ago. The problem is that ReceivedInv and RequestedTx cause the "PostGetRequestable" invariants to be violated until the next SetTimePoint. That's currently not a problem as the only way to observe that is through GetRequestable, which calls SetTimePoint. But if SetTimePoint is made a separate public function, we'd need to either:
* Add a rule that GetRequestable can only be called after SetTimePoint, with no mutators in between - effectively forcing the API to do the same thing it's doing now, but with the burden on the caller.
* Add more state, remembering whether anything may have broken the invariants, and refusing GetRequestable in that condition.
* Improve even more structure, and precompute the result of all possible GetRequestable (across all peers), and cache them inside TxRequestTracker, to be fetched/deleted whenever GetRequestable is called. This would significantly worsen worst-case memory usage, I think.
* Add the ability to install a callback function for expirations, instead of returning them from GetRequestable.

So all things considered, I think this is a decent approach. The callback idea seems reasonable, and perhaps can be extended to more than just expirations, though if that's what we pick, I'd like to keep it as a follow-up.

Reply:
> ReceivedInv and RequestedTx cause the "PostGetRequestable" invariants to be violated until the next SetTimePoint.

How about calling `PromoteCandidateReady()` from `ReceivedInv()` if `reqtime` is before `m_time` and calling `MakeCompleted()` from `RequestedTx()` if `expiry` is before `m_time`? That means that the invariants in `PostGetRequestableSanityCheck()` are actually invariant all the time. Now that `PostGetRequestableSanityCheck()` doesn't take a time argument, it's only acting on internal `TxRequestTracker` data, so as long as all public methods maintain the invariants, it should be safe to call any time (and perhaps could be combined into `SanityCheck()`)

Reply:
> The callback idea seems reasonable, and perhaps can be extended to more than just expirations

Yes, this could potentially be useful if there are other events that we want to expose to net_prociessing.

> I'd like to keep it as a follow-up.

I think whatever we do should be kept as a follow-up. Time to freeze this PR and get it merged :)

Reply:
Agree.

Reply:
We could use clang's lock annotations for the additional rule so the extra burden is on the compiler rather than the caller? Have SetTimePoint return an RAII mutex-like thing, make GetRequestable require the mutex-like is held, and make the other functions require that it not be held. cf #18017

Reply:
I don't like the idea of exporting more state and conditions to the caller, but we can discuss in the follow-up PR. Marking this resolved.